OSDOCS-9764: adds GitOps assembly MicroShift (#73201)
openshift/openshift-docs: microshift_running_apps/microshift-gitops.adoc

:_mod-docs-content-type: ASSEMBLY
[id="microshift-gitops"]
= Automating application management with the GitOps controller
include::_attributes/attributes-microshift.adoc[]
:context: microshift-gitops

toc::[]

> GitOps with Argo CD for {microshift-short} is a lightweight, optional add-on controller derived from the Red Hat OpenShift GitOps Operator. GitOps for {microshift-short} uses the command-line interface (CLI) of Argo CD to interact with the GitOps controller that acts as the declarative GitOps engine. You can can consistently configure and deploy Kubernetes-based infrastructure and applications across clusters and development lifecycles.

Reviewer: ```suggestion
GitOps with Argo CD for {microshift-short} is a lightweight, optional add-on controller derived from the Red Hat OpenShift GitOps Operator. GitOps for {microshift-short} uses the command-line interface (CLI) of Argo CD to interact with the GitOps controller that acts as the declarative GitOps engine. You can consistently configure and deploy Kubernetes-based infrastructure and applications across clusters and development lifecycles.
```

[id="microshift-gitops-can-do_{context}"]
== What you can do with the GitOps agent

> By using the GitOps with Argo CD agent with {microshift-short}, you can leverage the following principles and tasks:

Reviewer: ```suggestion
By using the GitOps with Argo CD agent with {microshift-short}, you can leverage the following principles and tasks:
```

Reviewer: Vale complains about "leverage", but I read the rest of the text and honestly, I can't come up with anything better than "leverage". Maybe "take advantage of"???

Author: Yea, "leverage" is very markety. I went with "utilize" and deleted "and tasks." It really is Git principles.

** For example, if you have 1,000 edge devices, each using {microshift-short} and a local GitOps agent, you can easily add or update an application on all 1,000 devices with just one change in your central Git repository.
* The Git repository contains a declarative description of the infrastructure you need in your specified environment and contains an automated process to make your environment match the described state.
* You can also use the Git repository as an audit trail of changes so that you can to create processes based on Git flows such as review and approval for merging pull requests that implement configuration changes.

[id="microshift-gitops-cannot-do_{context}"]
== Limitations of using the GitOps agent with {microshift-short}
GitOps with Argo CD for {microshift-short} has the following differences from the Red Hat OpenShift GitOps Operator:

* The `gitops-operator` component is not used with {microshift-short}.
* To maintain the small resource use of {microshift-short}, the Argo CD web console is not available. You can use the Argo CD CLI or use a pull-based approach.

> * Because {microshift-short} is single-node, there is no multi-cluster support. Each instance of {microshift-short} is paired with a local GitOps agent.

Reviewer 2: no, this actually means multi-cluster. the openshift-gitops operator on full blown OCP is frequently being used to manage application on other clusters (that is push-mode then, as the argo pushes out changes to other clusters). So this "multi-cluster" is  correct here.

> * Use sos reports for debugging because `oc admn must-gather` is not available in {microshift-short}.

Author: Looks like we're both wrong, it's `oc adm must-gather` https://docs.openshift.com/container-platform/4.15/cli_reference/openshift_cli/administrator-cli-commands.html#oc-adm-must-gather Fixing!

:_mod-docs-content-type: ASSEMBLY
[id="microshift-gitops"]

> = Automating application management with the GitOps agent

Reviewer 3: Should we mention `controller` instead of `agent` ?

include::_attributes/attributes-microshift.adoc[]
:context: microshift-gitops

toc::[]

{microshift-short} GitOps is a lightweight, optional add-on agent derived from the Red Hat OpenShift GitOps Operator. GitOps for {microshift-short} uses the command-line interface (CLI) of Argo CD as the declarative GitOps engine. You can can consistently configure and deploy Kubernetes-based infrastructure and applications across clusters and development lifecycles.

[id="microshift-gitops-can-do_{context}"]
== What you can do with the GitOps agent
By using the GitOps agent with {microshift-short}, you can leverage the following principles and tasks:

> * Extend the core principles of developing and maintaining software in a Git repository to the creation and management of your cluster and application configuration files. This means you can update the single repository and GitOps automates the deployment of new applications or updates to existing ones.

Reviewer 4: I might not be understanding that sentence right so ignore if Im wrong 
```suggestion
* Extend the core principles of developing and maintaining software in a Git repository. Including the creation and management of your cluster as well as the application configuration files. With this add on, you can update the single repository and GitOps automates the deployment of new applications or updates to existing ones.
```

> * Create and manage your clusters and application configuration files using the core principles of developing and maintaining software in a Git repository. Using GitOps with ArgoCD you can update the single repository and GitOps automates the deployment of new applications or updates to existing ones.

Reviewer 5: ```suggestion
* Create and manage your clusters and application configuration files using the core principles of developing and maintaining software in a Git repository. You can update the single repository and GitOps automates the deployment of new applications or updates to existing ones.
```

"Using GitOps with ArgoCD" looks redundant as you have mentioned that part in the lead sentence.

> ** For example, if you have 1,000 edge devices, each using {microshift-short} and a local GitOps agent, you can easily add or update an application on all 1,000 devices with just one change in your central Git repository.

Reviewer 4: I don't think these needs to be in a sub bullet, I think just a space would be good 
```suggestion

For example, if you have 1,000 edge devices, each using {microshift-short} and a local GitOps agent, you can easily add or update an application on all 1,000 devices with just one change in your central Git repository.
```

Reviewer 3: Should we also mention that the changes to the Git Repository are controlled and governed by the PR approval process and all changes could be audited, (like who edited, who approved etc).

* The Git repository contains a declarative description of the infrastructure you need in your specified environment and contains an automated process to make your environment match the described state.
** For example,
//TODO: what can we do? set memory limits or template over networking settings, etc?

> * The repository also contains the entire state of the system so that the trail of changes to the system state are visible and auditable.

Reviewer 2: I would remove that sentence, as this is not true. It's not the entire state, as e.g. the state of the os and microshift itself is not managed by gitops. Its really focusing on k8s level. Maybe:
Utilise the git repository as an audit trail of changes, and to easily impliement approval / review processes based on git flows like merging of pull requests for a config change.

> //TODO: what's the most important value prop of this visibility for microshift customers?

Reviewer 2: see above: Application Lifecycle Management - dynamically manage Application / Config of solutions with a gitops approach.

> //Do we have any infrastructure services?

Reviewer 2: no

[id="microshift-gitops-cannot-do_{context}"]
== Limitations of using the GitOps agent with {microshift-short}
GitOps with Argo CD for {microshift-short} has the following differences from the Red Hat OpenShift GitOps Operator:

> * To maintain the small resource use of {microshift-short}, the ArgoCD console is not available. All interactions must use the CLI.

Reviewer 3: There is no `gitops-operator` component as well.

* All interactions must use the Argo CD CLI. To maintain the small resource use of {microshift-short}, the Argo CD web console is not available.
* Because {microshift-short} is single-node, there is no multi-cluster support. Each instance of {microshift-short} is paired with a local GitOps agent.
//* Debugging is provided by...

> //How do users debug GitOps with MicroShift, since MicroShift does not support oc adm must-gather

Reviewer 2: I think we can simply point to  microshift sosreports for troubleshooting - those should include the high level infos like pod states.

:_mod-docs-content-type: ASSEMBLY
[id="microshift-gitops"]
= Automating application management with the GitOps controller
include::_attributes/attributes-microshift.adoc[]
:context: microshift-gitops

toc::[]

GitOps with Argo CD for {microshift-short} is a lightweight, optional add-on controller derived from the Red Hat OpenShift GitOps Operator. GitOps for {microshift-short} uses the command-line interface (CLI) of Argo CD to interact with the GitOps controller that acts as the declarative GitOps engine. You can can consistently configure and deploy Kubernetes-based infrastructure and applications across clusters and development lifecycles.

> 

Reviewer 6: We need a section here on how to install gitops RPM.
```
$ sudo subscription-manager repos --enable=gitops-1.12-for-rhel-9-$(uname -m)-rpms
$ sudo dnf install microshift-gitops
```

Reviewer 6: We should probably combine these PRs.
Anyway, I commented in the other PR as well

Reviewer 2: the release-info package is only required for ostree based airgapped/offline deployments when the container images required by gitops should be added to the ostree image.

Reviewer 2: Actually, the CLI is not **required** to be used with MicroShift. Better wording:
microshift-gitops can be used in a pull-based gitops approach: each edge device runs a small local git tops controled (based on Argo). That argo instance is reconciling a config from a remote git repo with the local API. It acts as the declarative GitOps engine. You can can consistently configure and deploy Kubernetes-based infrastructure and applications.

** For example, if you have 1,000 edge devices, each using {microshift-short} and a local GitOps agent, you can easily add or update an application on all 1,000 devices with just one change in your central Git repository.
* The Git repository contains a declarative description of the infrastructure you need in your specified environment and contains an automated process to make your environment match the described state.
** For example,
//TODO: what can we do? set memory limits or template over networking settings, etc?
* The repository also contains the entire state of the system so that the trail of changes to the system state are visible and auditable.
//TODO: what's the most important value prop of this visibility for microshift customers?
//Do we have any infrastructure services?

[id="microshift-gitops-cannot-do_{context}"]
== Limitations of using the GitOps agent with {microshift-short}
GitOps with Argo CD for {microshift-short} has the following differences from the Red Hat OpenShift GitOps Operator:

* The `gitops-operator` component is not used with {microshift-short}.

> * All interactions must use the CLI. To maintain the small resource use of {microshift-short}, the ArgoCD console is not available.

Reviewer 7: ```suggestion
* All interactions must use the CLI. To maintain the small resource use of {microshift-short}, the Argo CD web console is not available.
```

> * All interactions must use the CLI. To maintain the small resource use of {microshift-short}, the Argo CD web console is not available.

Reviewer 8: ```suggestion
* All interactions must use the Argo CD CLI. To maintain the small resource use of {microshift-short}, the Argo CD web console is not available.
```

* Because {microshift-short} is single-node, there is no multi-cluster support. Each instance of {microshift-short} is paired with a local GitOps agent.
//* Debugging is provided by...
//How do users debug GitOps with MicroShift, since MicroShift does not support oc adm must-gather

[id="additional-resources_microshift-gitops_{context}"]
[role="_additional-resources"]
== Additional resources

> * link:https://access.redhat.com/documentation/en-us/red_hat_openshift_gitops/1.11[Red Hat OpenShift GitOps]

Reviewer 6: ```suggestion
* link:https://access.redhat.com/documentation/en-us/red_hat_openshift_gitops/1.12[Red Hat OpenShift GitOps]
```
